Sources/SwiftDocC/Infrastructure/Link Resolution/PathHierarchy+Find.swift

```diff
@@ -355,29 +355,62 @@ extension PathHierarchy {
+            return Error.unknownName(
                 partialResult: (
                     node,
                     pathForError(of: rawPathForError, droppingLast: remaining.count)
                 ),
                 remaining: Array(remaining),
-                candidates: disambiguationTree.disambiguatedValues().map {
-                    (node: $0.value, disambiguation: String($0.disambiguation.makeSuffix().dropFirst()))
-                }
+                availableChildren: Set(node.children.keys)
             )
         }
-        
-        return Error.unknownName(
+
+        // Use a empty disambiguation suffix for the preferred symbol, if there
+        // is one, which will trigger the warning to suggest removing the
+        // suffix entirely.
+        let candidates = disambiguationTree.disambiguatedValues()
+        let favoredSuffix = favoredSuffix(from: candidates)
+        let suffixes = candidates.map { suffix(for: $0.disambiguation) }
+        let candidatesAndSuffixes = zip(candidates, suffixes).map { (candidate, suffix) in
+            if suffix == favoredSuffix {
+                return (node: candidate.value, disambiguation: "")
+            } else {
+                return (node: candidate.value, disambiguation: suffix)
+            }
+        }
+        return Error.unknownDisambiguation(
             partialResult: (
                 node,
                 pathForError(of: rawPathForError, droppingLast: remaining.count)
             ),
             remaining: Array(remaining),
-            availableChildren: Set(node.children.keys)
+            candidates: candidatesAndSuffixes
         )
     }
-    
+
+    /// Check if exactly one of the given candidate symbols is preferred, because it is not disfavored
+    /// for link resolution and all the other symbols are.
+    /// - Parameters:
+    ///   - from: An array of candidate node and disambiguation tuples.
+    /// - Returns: An optional string set to the disambiguation suffix string, without the hyphen separator e.g. "abc123",
+    ///            or nil if there is no preferred symbol.
+    private func favoredSuffix(from candidates: [(value: PathHierarchy.Node, disambiguation: PathHierarchy.DisambiguationContainer.Disambiguation)]) -> String? {
+        guard let favored = candidates.singleMatch({
+            !$0.value.specialBehaviors.contains(PathHierarchy.Node.SpecialBehaviors.disfavorInLinkCollision)
+        }) else { return nil }
+        return suffix(for: favored.disambiguation)
+    }
+
+    /// Return the suffix string "abc123" for the given disambiguation, without the hyphen separator
+    /// - Parameters:
+    ///   - from: A PathHierarchy disambiguation structure
+    /// - Returns: The corresponding disambiguation suffix string, without the hyphen separator e.g. "abc123"
+    private func suffix(for disambiguation: PathHierarchy.DisambiguationContainer.Disambiguation) -> String {
```

Comment:
This function doesn't access any path hierarchy information, so it could be a free function. However, since it only operates on the only argument that it is passed, it's more idiomatic in Swift to use extend that type and call it as `value.something()` instead of `something(value)`.

That said, `disambiguation` already has a `makeSuffix` function. Do we need both? It seems like the only reason to remove the hyphen is to present the suffix without the hash in the solution message, while adding back the hype for the replacement. 

I think we can remove this function, pass the suffix with the hyphen, and drop the hyphen in `replacementOperationDescription(from:to:)` instead.

Reply:
Ok I've simplified this even more and got rid of the `suffix` function, just calling `makeSuffix` directly. I didn't make any changes to `replacementOperationDescription` since that is also used in a number of other code paths.